### PR TITLE
[feat] [client] Enable custom Encrypt/Decrypt Methods for ReaderImplementation.

### PR DESCRIPTION
**Pull request by shashrath**
### Motivation
Currently, when we are creating a reader, we don't have the option to pass our custom MessageCrypto.
Suppose we are using KMS for encrypting and decrypting we can pass our custom MessageCrypto instance to the reader now with these changes.

### Modifications

Added the MessageCrypto field to the ReaderConfigurationData.
Added the setter and getting methods into the ReaderImpl class.

### Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

### Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API: (yes)
  - The schema: (no)
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: (no)

### Documentation

- [x] `doc-required` 
Doc changes might be required.

### CODE REVIEW

**Comment by Anonymitaet** [comment]
Hi @shashrath 

I see you've added explanations in Java client API doc.

1. Does this feature belong to [End-to-end Encryption of reader](https://docs.google.com/spreadsheets/d/1YHYTkIXR8-Ql103u-IMI18TXLlGStK8uJjDsOOA0T20/edit#gid=1784579914&range=98:98)? 

2. If so, will you consider adding docs [here](https://pulsar.apache.org/docs/en/next/client-libraries-java/)? Like create a section (`End-to-end encryption`) and add explanations there?

**Review by Jason918 — COMMENTED** [review]
LGTM. But it's better to add some unit test to verify this.

**Review by codelipenghui — COMMENTED** [review]
The change looks good to me, could you please add a unit test for covering the new changes?

**Comment by Anonymitaet** [comment]
> Hi @shashrath
> 
> I see you've added explanations in Java client API doc.
> 
> 1. Does this feature belong to [End-to-end Encryption of reader](https://docs.google.com/spreadsheets/d/1YHYTkIXR8-Ql103u-IMI18TXLlGStK8uJjDsOOA0T20/edit#gid=1784579914&range=98:98)?
> 2. If so, will you consider adding docs [here](https://pulsar.apache.org/docs/en/next/client-libraries-java/)? Like create a section (`End-to-end encryption`) and add explanations there?

Hi @shashrath any feedback on this?

**Comment by shashrath** [comment]
> > Hi @shashrath
> > I see you've added explanations in Java client API doc.
> > 
> > 1. Does this feature belong to [End-to-end Encryption of reader](https://docs.google.com/spreadsheets/d/1YHYTkIXR8-Ql103u-IMI18TXLlGStK8uJjDsOOA0T20/edit#gid=1784579914&range=98:98)?
> > 2. If so, will you consider adding docs [here](https://pulsar.apache.org/docs/en/next/client-libraries-java/)? Like create a section (`End-to-end encryption`) and add explanations there?
> 
> Hi @shashrath any feedback on this?

Hi @Anonymitaet yeah this feature comes under End-to-end Encryption/Decryption of reader.
Yeah will try to update the docs there.

**Comment by Anonymitaet** [comment]
@shashrath feel free to ping me if you need a doc review.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsReaderImpl.java [hunk 1]
```diff
@@ -94,6 +94,10 @@ public void reachedEndOfTopic(Consumer<T> consumer) {
         if (readerConfiguration.getCryptoKeyReader() != null) {
             consumerConfiguration.setCryptoKeyReader(readerConfiguration.getCryptoKeyReader());
         }
+
+        if(readerConfiguration.getMessageCrypto() != null){
```
Comment (eolivelli) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MultiTopicsReaderImpl.java:98]: Nit: missing ehitespaces

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ReaderImpl.java [hunk 2]
```diff
@@ -106,6 +106,10 @@ public void reachedEndOfTopic(Consumer<T> consumer) {
             consumerConfiguration.setCryptoKeyReader(readerConfiguration.getCryptoKeyReader());
         }
 
+        if(readerConfiguration.getMessageCrypto() != null){
```
Comment (eolivelli) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ReaderImpl.java:109]: Nit: missing whitespaces

**Review by eolivelli — CHANGES_REQUESTED** [review]
This patch is very useful, can you please add unit tests?
If you have time please also add a paragraph to docs, if you can't we can open a ticket.

Feel free to ask for help, your contribution is very nice

**Comment by poorbarcode** [comment]
Since we will start the RC version of `3.0.0` on `2023-04-11`, I will change the label/milestone of PR who have not been merged.
- The PR of type `feature` is deferred to `3.1.0`
- The PR of type `fix` is deferred to `3.0.1`

So drag this PR to `3.1.0`

**Review by tisonkun — COMMENTED** [review]
LGTM. I did it for producer builder in https://github.com/apache/pulsar/pull/19939.

Let me try it pick up this PR.

**Review by tisonkun — APPROVED** [review]

**Review by eolivelli — APPROVED** [review]
+1

**Review by Jason918 — APPROVED** [review]
LGTM

**Comment by tisonkun** [comment]
Thanks for your review.

Merging...

**Comment by Anonymitaet** [comment]
> > > Hi @shashrath
> > > I see you've added explanations in Java client API doc.
> > > 
> > > 1. Does this feature belong to [End-to-end Encryption of reader](https://docs.google.com/spreadsheets/d/1YHYTkIXR8-Ql103u-IMI18TXLlGStK8uJjDsOOA0T20/edit#gid=1784579914&range=98:98)?
> > > 2. If so, will you consider adding docs [here](https://pulsar.apache.org/docs/en/next/client-libraries-java/)? Like create a section (`End-to-end encryption`) and add explanations there?
> > 
> > 
> > Hi @shashrath any feedback on this?
> 
> Hi @Anonymitaet yeah this feature comes under End-to-end Encryption/Decryption of reader. Yeah will try to update the docs there.

Hi @shashrath any progress on the doc? Thanks!

**Pull request merged**